controllers/machine_controller_phases.go

```diff
@@ -229,6 +229,11 @@ func (r *MachineReconciler) reconcileInfrastructure(ctx context.Context, cluster
 		return nil
 	}
 
+	// Pull the failureDomain from the infrastructure provider if it is not already set.
+	if m.Spec.FailureDomain == nil {
```

Comment:
Is this logic the same as https://github.com/kubernetes-sigs/cluster-api/pull/2629/files#diff-c4709802144ce7e246425de030409a0eR263-R272 ?

Reply:
slightly different, the linked code implements pulling the failureDomain from the infraConfig using unstructured, this just gates calling that to set m.Spec.FailureDomain only if m.Spec.FailureDomain is not already set.

Reply:
i guess my question is more like: should we do the conditional check down there rather than here? or just keep this one and remove lines 263-272?

Reply:
I'll take another look in the morning and see if I can clean it up a bit. I did want to make sure that it was testable a bit more independently from the rest of the phase, but I likely introduced a bug in the handling as @ncdc pointed out.